### PR TITLE
DirectMethod Invocation sample

### PR DESCRIPTION
**Pull request by varunpuranik**
This is a sample that contains 2 V2 modules that shows how Direct method invocation works. 
One module generates data, and the other module invokes methods on the first module based on some simple conditions. 
This also demonstrates how to create and use the ServiceClient in the module code to invoke the direct methods. Adding this sample as there is currently not much existing documentation that shows how to do this.

### CODE REVIEW

File: v2/samples/direct_method_invocation_sample/deployment.json [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+          }
+        },
+        "modules": {
+          "engineAnalyzer": {
+            "version": "1.0",
+            "type": "docker",
+            "status": "running",
+            "restartPolicy": "always",
+            "settings": {
+              "image": "edgetest.azurecr.io/engineanayzer:varun2",
```
Comment (avranju) [inline v2/samples/direct_method_invocation_sample/deployment.json]: Can we have this image be hosted in Docker hub? Or alternately, put some placeholder image name here that users are expected to replace with an actual image reference.

File: v2/samples/direct_method_invocation_sample/AnalyzerModule/AnalyzerModule.csproj [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|netcoreapp2.0|AnyCPU'">
+    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
+    <TreatSpecificWarningsAsErrors />
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Azure.Devices.Client" Version="1.6.0-preview-001" />
+    <PackageReference Include="Microsoft.Azure.Devices" Version="1.5.0-preview-001" />
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.0.0-preview2-final" />
```
Comment (avranju) [inline v2/samples/direct_method_invocation_sample/AnalyzerModule/AnalyzerModule.csproj]: We can probably switch to using the non-preview 2.0 version for all the `Microsoft.Extensions.*` assemblies.

File: v2/samples/direct_method_invocation_sample/deployment.json (hunk 1 above)
Comment (varunpuranik) [reply v2/samples/direct_method_invocation_sample/deployment.json]: Ah, thanks for catching that. That was an oversight.. meant to use a placeholder there. Will fix.

File: v2/samples/direct_method_invocation_sample/AnalyzerModule/AnalyzerModule.csproj (hunk 2 above)
Comment (varunpuranik) [reply v2/samples/direct_method_invocation_sample/AnalyzerModule/AnalyzerModule.csproj]: Sure..

File: v2/samples/direct_method_invocation_sample/README.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+--------------
+
+A module can invoke a direct method on another module or a downstream device (i.e. a device
+connected to the same EdgeHub), by using its EdgeHubConnectionString. It cannot invoke direct methods
+on modules on other edge devices, or devices not connected to the same EdgeHub (unless it uses an 
+IotHub scoped connection string).
+
+Here are the steps to invoke a direct method from one module to another -
+
+- You need to use the ServiceClient to invoke direct methods from a module. The ServiceClient is in the Microsoft.Azure.Devices nuget package which is not added by default, so you need to manually add that. Make sure to add a reference to a preview* version of the nuget package, to get modules support.
```
Comment (jonbgallant) [inline v2/samples/direct_method_invocation_sample/README.md]: Can you include a CLI statement that shows exactly what you are asking them to do here?

File: v2/samples/direct_method_invocation_sample/README.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+A module can invoke a direct method on another module or a downstream device (i.e. a device
+connected to the same EdgeHub), by using its EdgeHubConnectionString. It cannot invoke direct methods
+on modules on other edge devices, or devices not connected to the same EdgeHub (unless it uses an 
+IotHub scoped connection string).
+
+Here are the steps to invoke a direct method from one module to another -
+
+- You need to use the ServiceClient to invoke direct methods from a module. The ServiceClient is in the Microsoft.Azure.Devices nuget package which is not added by default, so you need to manually add that. Make sure to add a reference to a preview* version of the nuget package, to get modules support. 
+
+- Create a ServiceClient using the EdgeHubConnectionString that is passed in to the module as an
```
Comment (jonbgallant) [inline v2/samples/direct_method_invocation_sample/README.md]: into

File: v2/samples/direct_method_invocation_sample/README.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+- Invoke direct method on another module like this -
+```
+var response = await serviceClient.InvokeDeviceMethodAsync(message.ConnectionDeviceId, message.ConnectionModuleId, new CloudToDeviceMethod("Reset"));                    
+```
+> Note: An incoming message contains the device Id in the property ConnectionDeviceId, and the
+module Id in the property ConnectionModuleId (if the message is from another module). You can 
+use those values to invoke a method on that device or module.
+
+- Receiving direct method invocations from a module is same as if receiving it from the cloud.
```
Comment (jonbgallant) [inline v2/samples/direct_method_invocation_sample/README.md]: For completeness, please include what that looks like here, for folks who haven't seen it yet.

File: v2/samples/direct_method_invocation_sample/AnalyzerModule/.gitignore [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# .NET Core
```
Comment (jonbgallant) [inline v2/samples/direct_method_invocation_sample/AnalyzerModule/.gitignore]: Can you rename the project to EngineAnalyzer so it is consist with the docker image name?

File: v2/samples/direct_method_invocation_sample/EngineSimulatorModule/.gitignore [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# .NET Core
```
Comment (jonbgallant) [inline v2/samples/direct_method_invocation_sample/EngineSimulatorModule/.gitignore]: Can you rename the project to EngineSimulator so it is consist with the docker image name?

File: v2/samples/direct_method_invocation_sample/AnalyzerModule/.gitignore (hunk 6 above)
Comment (jonbgallant) [reply v2/samples/direct_method_invocation_sample/AnalyzerModule/.gitignore]: It's a personal preference, but I like to have the module "key" be the same as the docker image name and the routes match as well:

image

File: v2/samples/direct_method_invocation_sample/AnalyzerModule/Docker/linux-x64/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,11 @@
+FROM microsoft/dotnet:2.0.0-runtime
```
Comment (darobs) [inline v2/samples/direct_method_invocation_sample/AnalyzerModule/Docker/linux-x64/Dockerfile]: Shall we bum up to 2.0.5 runtime?

Comment (myagley) [reply v2/samples/direct_method_invocation_sample/AnalyzerModule/Docker/linux-x64/Dockerfile]: 👍

**Comment by myagley** [comment]
Thanks for doing this Varun!

File: v2/samples/direct_method_invocation_sample/README.md (hunk 3 above)
Comment (varunpuranik) [reply v2/samples/direct_method_invocation_sample/README.md]: Sure..

File: v2/samples/direct_method_invocation_sample/README.md (hunk 5 above)
Comment (varunpuranik) [reply v2/samples/direct_method_invocation_sample/README.md]: Sure, makes sense.

**Pull request merged**